### PR TITLE
Sameer/allowlist minter

### PR DESCRIPTION
**Pull request by crazzywizard**
### Purpose
* created allowlist minter contract
* created public minter contract
* created shared functionality
* added new tests

### CODE REVIEW

File: src/minter/PublicMinter.sol [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        uint256 pfpTokenId = ICre8ors(cre8orsNFT).adminMint(
+            recipient,
+            quantity
+        );
+        payable(address(cre8orsNFT)).call{value: msg.value}("");
+        _lockUp(carts, pfpTokenId - quantity + 1);
+        return pfpTokenId;
+    }
+
+    modifier inactiveSale(address recipient) {
```
Comment (SweetmanTech) [inline src/minter/PublicMinter.sol:56]: how can we rename this modifier to be more descriptive?
```suggestion
    modifier onlyPublicSaleOrAlreadyMinted(address recipient) {
```

File: src/minter/PublicMinter.sol [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        /**  @dev This is the only change from AllowlistMinter
+         * This is so that anyone with a
+         *                 pfp can mint from the public sale
+         *                 which before public sale being active
+         *                 should be only discount/passport holders
+         */
+        if (
+            !ICre8ors(cre8orsNFT).saleDetails().publicSaleActive &&
+            ICre8ors(cre8orsNFT).mintedPerAddress(recipient).totalMints == 0
```
Comment (SweetmanTech) [inline src/minter/PublicMinter.sol:65]: have you written a unit test to verify:
- `A` uses Passport to mint on 8/8
- `A` unlocks passport & sends to B
- `B` tries to mint on PublicMinter
- - expect `revert`

File: src/utils/SharedPaidMinterFunctions.sol [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import {ICre8ors} from "../interfaces/ICre8ors.sol";
+
+contract SharedPaidMinterFunctions {
+    address public cre8orsNFT;
+    address public minterUtility;
+    modifier verifyCost(IMinterUtilities.Cart[] memory carts) {
+        uint256 totalCost = IMinterUtilities(minterUtility).calculateTotalCost(
+            carts
+        );
+        require(msg.value >= totalCost, "Not enough ETH sent");
```
Comment (SweetmanTech) [inline src/utils/SharedPaidMinterFunctions.sol:14]: check should be the opposite right?

```suggestion
        require(msg.value < totalCost, "Not enough ETH sent");
```

Comment (SweetmanTech) [reply src/utils/SharedPaidMinterFunctions.sol:14]: also, save gas with `rever` and custom error
```suggestion
if (msg.value < totalCost) {
  revert InsufficientCustomError()
}
```

File: src/utils/SharedPaidMinterFunctions.sol [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        _;
+    }
+
+    function _lockUp(
+        IMinterUtilities.Cart[] memory carts,
+        uint256 startingTokenId
+    ) internal {
+        ILockup lockup = ICre8ors(cre8orsNFT).lockup();
```
Comment (SweetmanTech) [inline src/utils/SharedPaidMinterFunctions.sol:22]: after byanca changes, I think `lockup` can only be referenced from `cre8ing` 💭

Comment (SweetmanTech) [reply src/utils/SharedPaidMinterFunctions.sol:22]: `git pull origin main`
`forge test`

File: src/utils/SharedPaidMinterFunctions.sol [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        } else if (tier == 2) {
+            lockupDate = block.timestamp + tier2.lockup;
+            tierPrice = tier2.price;
+        } else {
+            lockupDate = block.timestamp + tier3.lockup;
+            tierPrice = tier3.price;
+        }
+        return abi.encode(lockupDate, tierPrice);
+    }
+}
```
Comment (SweetmanTech) [inline src/utils/SharedPaidMinterFunctions.sol:72]: suggestion by GPT to improve this fx

```
function _getLockUpDateAndPrice(uint256 tier) internal view returns (bytes memory) {
    require(tier >= 1 && tier <= 3, "Invalid tier");

    IMinterUtilities.TierInfo[] memory tiers = abi.decode(
        IMinterUtilities(minterUtility).getTierInfo(),
        (IMinterUtilities.TierInfo[])
    );

    IMinterUtilities.TierInfo memory selectedTier = tiers[tier - 1];
    uint256 lockupDate = block.timestamp + selectedTier.lockup;
    uint256 tierPrice = selectedTier.price;

    return abi.encode(lockupDate, tierPrice);
}
```

Screenshot 2023-08-01 at 1 43 18 PM
Screenshot 2023-08-01 at 1 43 50 PM

File: src/minter/PublicMinter.sol (hunk 2 above)
Comment (crazzywizard) [reply src/minter/PublicMinter.sol:65]: yes I have, it does revert as expected.

File: src/utils/SharedPaidMinterFunctions.sol (hunk 4 above)
Comment (crazzywizard) [reply src/utils/SharedPaidMinterFunctions.sol:22]: done!

File: src/minter/AllowlistMinter.sol [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import {ICre8ors} from "../interfaces/ICre8ors.sol";
+import {ILockup} from "../interfaces/ILockup.sol";
+import {IERC721A} from "lib/ERC721A/contracts/interfaces/IERC721A.sol";
+import {IERC721Drop} from "../interfaces/IERC721Drop.sol";
+import {MerkleProof} from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import {IMinterUtilities} from "../interfaces/IMinterUtilities.sol";
+import {SharedPaidMinterFunctions} from "../utils/SharedPaidMinterFunctions.sol";
+
+contract AllowlistMinter is SharedPaidMinterFunctions {
+    constructor(address _cre8orsNFT, address _minterUtility) {
```
Comment (SweetmanTech) [inline src/minter/AllowlistMinter.sol:14]: nit, shouldn't need to store `cre8orsNFT` in storage.

File: src/utils/MinterUtilities.sol [hunk 7]
```diff
@@ -69,12 +72,13 @@ contract MinterUtilities is IMinterUtilities {
...
         uint256 maxQuantity = maxAllowedQuantity(totalClaimed);
-        uint256 quantityRemaining = maxQuantity - totalMints;
+
+        int256 quantityRemaining = int256(maxQuantity) - int256(totalMints);
 
         if (quantityRemaining < 0) {
             return 0;
         }
-        return quantityRemaining;
+        return uint256(quantityRemaining);
```
Comment (SweetmanTech) [inline src/utils/MinterUtilities.sol:81]: this accomplishes the same goal without wrapping in `int256` right?

```
 if (maxQuantity < totalMints) {
            return 0;
  }
return uint256(maxQuantity - totalMints);
```

Comment (crazzywizard) [reply src/utils/MinterUtilities.sol]: yep it does, implemented your suggestion

**Review by SweetmanTech — APPROVED** [review]
ship iit 🚢

**Pull request merged**